### PR TITLE
Add av1 encode Interfaces

### PR DESCRIPTION
**Pull request by Xuhenry1**
Add av1 encode DDI and introduce corresponding attributes.

### CODE REVIEW

**Comment by XinfengZhang** [comment]
please also check the document and github pages, refer #532

File: va/va_enc_av1.h
Comment (feiwan1) [inline va/va_enc_av1.h:834]: inserted -> overwritten may be better?

**Comment by Xuhenry1** [comment]
> please also check the document and github pages, refer #532

thanks, updated

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:760]: Updated 3: Invalid

File: va/va_str.c
Comment (XinfengZhang) [reply va/va_str.c:149]: no new buffer type added.

Comment (Xuhenry1) [reply va/va_str.c:149]: Yes, no new buffer type add, reuse the same buffertype and add traces func for those buffer.

**Comment by xhaihao** [comment]
There are lots of compiling errors, e.g.

../va/va_trace.c:4176:86: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’
 4176 |     va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_enabled = %d\n", p->segments.bits.segmentation_enabled);
      |                                                                                      ^
../va/va_trace.c:4177:89: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’
 4177 |     va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_update_map = %d\n", p->segments.bits.segmentation_update_map);
      |                                                                                         ^
../va/va_trace.c:4178:94: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’
 4178 |     va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_temporal_update = %d\n", p->segments.bits.segmentation_temporal_update);

**Comment by Xuhenry1** [comment]
> There are lots of compiling errors, e.g.
> 
> ../va/va_trace.c:4176:86: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’ 4176 | va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_enabled = %d\n", p->segments.bits.segmentation_enabled); | ^ ../va/va_trace.c:4177:89: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’ 4177 | va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_update_map = %d\n", p->segments.bits.segmentation_update_map); | ^ ../va/va_trace.c:4178:94: error: ‘VAEncSegParamAV1’ {aka ‘struct _VAEncSegParamAV1’} has no member named ‘bits’ 4178 | va_TraceMsg(trace_ctx, "\tsegments.bits.segmentation_temporal_update = %d\n", p->segments.bits.segmentation_temporal_update);

Thanks, compile issue resolved.

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:250]: what's tier? could not find the key words in spec.

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:255]: gop size is needed for encoder?

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:270]: so, when its CBR, it is target bitrate, max bit rete for VBR

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:250]: there will be seq_tier in spec, which specify the tier that the coded video sequence conforms to.  seq_tier will constrain MaxBitrate/MainCR at certain profile/level.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:255]: yes, it will let encoder know when to insert a intra_only frame, which could eliminate accumalted error.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:270]: yes, for VBR, the target bitrate will be generated through bits_per_second mutiplied target_percentage in #VAConfigAttribRateControl. 

Will add more comments for VBR.

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:339]: come from application?

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:361]: driver generate the segmentation_params and write back to headers? where the segmentation_bitoffset is defined. how encoder know the value.

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:366]: dont understand whether when the segment_number ==0, encoder will enable it or not. it looks different with previous description.

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:323]: is it SEG_LVL_MAX?

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:348]: missing segmentation_update_data?

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:570]: suppose it should be moved to behind order_hint, and it should be 3 bytes

File: va/va_enc_av1.h
Comment (XinfengZhang) [inline va/va_enc_av1.h:786]: reserved?

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:255]: yes, at driver side, need to know the Gop pic size to calculate per frame bitrate size for BRC mode.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:270]: currently we set target_percentage to 1 when there are no VAConfigAttribRateContro. so if it is VBR with no VAConfigAttribRateControl. VBR will be changed to CBR at driver side.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:339]: thx, changed to application.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:361]: bit_offset_segmentation is defined in pps, correct name and add cite here.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:366]: when segmentation_enabled == 1 and segment_number ==0,  driver will enable it with segment params generated by driver rather from application.

Comment (Xuhenry1) [reply va/va_enc_av1.h:366]: Add more detailed comments for explanation.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:348]: update data  will be transfered  through buffer VAEncPictureParameterBufferAV1.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:570]: thx, move location and use 3 bytes.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:786]: added.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:323]: yes

File: va/va_enc_av1.h
Comment (XinfengZhang) [reply va/va_enc_av1.h:271]: TBH, I think this is incorrect behavior, the behavior is not aligned with the definition.  but , other codec also defined this value. ...

**Review by XinfengZhang — APPROVED** [review]

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:324]: yes, just to align previous VA_AV1_MAXNUM_SEGMENTS. Agree it may be more clear if changed to SEG_LVL_MAX, which is aligned with spec.

File: va/va_enc_av1.h
Comment (Xuhenry1) [reply va/va_enc_av1.h:271]: Yes, This value seems redundant and only to be used when VAConfigAttribRateControl is not set at BRC mode.

File: va/va_trace.c
Comment (dvrogozh) [inline va/va_trace.c:4111]: There is a typo starting from this line: s/ctrl_l0/ctrl_l1/, in the argument (inside "" name is correct).

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: va/va_trace.c
Comment (Xuhenry1) [reply va/va_trace.c]: Thanks, corrected it.

**Review by dvrogozh — APPROVED** [review]
Looks good.

File: va/va_trace.c
Comment (feiwan1) [inline va/va_trace.c:4206]: Should be p->cdef_y/uv_strengths[i].
Others looks good to me.

Comment (Xuhenry1) [reply va/va_trace.c]: Thanks for reminder, updated.

**Review by feiwan1 — APPROVED** [review]
Looks good to me now.

**Comment by dvrogozh** [comment]
@Xuhenry1 : which changes were done after last approval by @feiwan1? Can you, please, make sure to post a note of what you have changed each time you push after approval, otherwise you invalidate the review status of the PR.

**Comment by Xuhenry1** [comment]
> @Xuhenry1 : which changes were done after last approval by @feiwan1? Can you, please, make sure to post a note of what you have changed each time you push after approval, otherwise you invalidate the review status of the PR.

yes, thanks, before approve by feiwan1, we remove one duplicate field in mode_control_flags.bits.reduced_tx_set.

**Pull request merged**
